pulpcore/cli/rpm/prune.py

```diff
@@ -0,0 +1,100 @@
+import gettext
+import typing as t
+
+import click
+from pulp_glue.common.context import EntityFieldDefinition, PulpException
+from pulp_glue.rpm.context import PulpRpmPruneContext, PulpRpmRepositoryContext
+
+from pulpcore.cli.common.generic import (
+    PulpCLIContext,
+    pass_pulp_context,
+    pulp_command,
+    resource_option,
+)
+
+_ = gettext.gettext
+
+multi_repository_option = resource_option(
+    "--repository",
+    "repositories",
+    default_plugin="rpm",
+    default_type="rpm",
+    context_table={"rpm:rpm": PulpRpmRepositoryContext},
+    multiple=True,
+    href_pattern=PulpRpmRepositoryContext.HREF_PATTERN,
+    help=_(
+        "Repository to prune, in the form '[[<plugin>:]<resource_type>:]<name>' or by href."
+        " Can be called multiple times."
+    ),
+)
+
+
+@pulp_command()
+@multi_repository_option
+@click.option(
+    "--all-repositories",
+    type=bool,
+    is_flag=True,
+    show_default=True,
+    default=False,
+    help=_("Prune *all* repositories accessible to the invoking user."),
+)
+@click.option(
+    "--keep-days",
+    type=int,
+    default=14,
+    help=_("Prune packages that were added to the specified repositories more than N days ago."),
+)
+@click.option(
+    "--dry-run",
+    type=bool,
+    is_flag=True,
+    show_default=True,
+    default=False,
+    help=_("Evaluate the prune-status of the specified repositories but DO NOT make any changes."),
+)
+@pass_pulp_context
+def prune_packages(
+    pulp_ctx: PulpCLIContext,
+    repositories: t.Iterable[EntityFieldDefinition],
+    all_repositories: t.Optional[bool],
+    keep_days: t.Optional[int],
+    dry_run: t.Optional[bool],
+) -> None:
+    """
+    Prune older Packages from the current-version of a repository/repositories.
+
+    Repositories can be specified by repeated --repository arguments.
+
+    "All" repositories can be specified by --all-repositories.
+
+    At least one repository, or --all-repositories, must be specified.
+
+    You may not specify --all-repositories *and* one or more specific repositories.
+    """
+    prune_ctx = PulpRpmPruneContext(pulp_ctx)
+    if not (all_repositories or repositories):
+        raise PulpException(
+            _("at least one --repository, or --all-repositories, must be specified")
+        )
+    elif all_repositories and repositories:
+        raise PulpException(
+            _("cannot specify --all-repositories and --repository at the same time")
+        )
+
+    if all_repositories:
+        repo_hrefs = ["*"]
+    else:
+        repo_hrefs = [
+            repository_ctx.pulp_href
+            for repository_ctx in repositories
+            if isinstance(repository_ctx, PulpRpmRepositoryContext)
+        ]
+
+    params = {
+        "repo_hrefs": repo_hrefs,
+        "keep_days": keep_days,
+        "dry_run": dry_run,
+    }
```

Comment:
Please try this:
```suggestion
    if all_repositories:
        repositories = ["*"]

    params = {
        "repo_hrefs": repositories,
        "keep_days": keep_days,
        "dry_run": dry_run,
    }
```

Reply:
I had tried that and it failed with "'repo_hrefs' is expected to be a list." I just did a little deeper experimentation,. and discovered that the problem is that, coming into this call, repositories is a generator. If I list(repositories), the correct validation happens in the correct place and All Is Well :)  next update will address.

Reply:
Nope - you can't "just pass" the generator, and you  can't pass `list(repositories)` because mypy complains `error: List comprehension has incompatible type List[str | PulpEntityContext | None]; expected List[str]` and the lint-check fails.

Restoring to original comprehension.

Reply:
Hmm. This raises the question whether it's good to allow `["*"]` as a "list of hrefs" on the server side. How is this represented in the openapi spec? If it were a "list of repository_hrefs" i'd expect to see `{"type": "array", "items": {"type": "str", "format": "uri"}`.

Reply:
What happens if you provided an empty list?
What, if you don't pass it at all?

The output of `pulp debug openapi operation --id rpm_prune_prune` or so should help us. The referenced (body) schema is of interest too.

Reply:
> What happens if you provided an empty list? What, if you don't pass it at all?
```
$ pulp rpm prune-packages 
Error: at least one --repository, or --all-repositories, must be specified
```

> The output of `pulp debug openapi operation --id rpm_prune_prune` or so should help us. The referenced (body) schema is of interest too.
```
$ pulp debug openapi schema --name PrunePackages  
{
  "type": "object",
  "description": "Serializer for prune-old-Packages operation.",
  "properties": {
    "repo_hrefs": {
      "type": "array",
      "items": {
        "type": "string",
        "minLength": 1
      },
      "description": "Will prune old packages from the specified list of repos. Use ['*'] to specify all repos. Will prune based on the specified repositories' latest_versions."
    },
    "keep_days": {
      "type": "integer",
      "format": "int64",
      "default": 14,
      "description": "Prune packages introduced *prior-to* this many days ago. Default is 14. A value of 0 implies 'keep latest package only.'",
      "minimum": 0
    },
    "dry_run": {
      "type": "boolean",
      "default": false,
      "description": "Determine what would-be-pruned and log the list of packages. Intended as a debugging aid."
    }
  },
  "required": [
    "repo_hrefs"
  ]
}
```

Reply:
So repo_hrefs is required and must contain at least one href or "*".
Thanks, that gives some insight.

Reply:
See current refactoring - took some pointers from https://github.com/pulp/pulp-cli/pull/749 , now simpler and keeps mypy happy.